### PR TITLE
Basic stats

### PR DESCRIPTION
**Pull request by RavenX8**
- Added level up and exp gain
- Added movement speed calculations
- Added attack speed calculations
- Updated item lua
- Fixed crash when picking up zuly
- Fixed windows compile issues
- Updated build system to be more robust

---
This change is [Reviewable](https://reviewable.io/reviews/dev-osrose/osirose-new/210)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -72,7 +90,7 @@ If you opted to keep them in place, the server binaries are located in the bin f
     $ ./bin/CharServer&
     $ ./bin/MapServer&
 
-After running for the first time, the servers will create a config file in the current directory.
+After running for the first time, the servers will create a config file in the current directory that you need to edit.
```
Comment (L3nn0x) [inline README.md:93]: We probably want to explain what each option does and how you can modify it. A help page in the github wiki?

File: src/map/src/utils/calculation.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    auto& values = entitySystem.get_component<Component::ComputedValues>(entity);
+    float moveSpeed = 0;
+
+    if (values.moveMode <= MoveMode::RUN) {
+      int itemSpeed = 65, itemNumber = 0;
+
+      if(entitySystem.has_component<Component::Inventory>(entity) == true) {
+        auto& inventory = entitySystem.get_component<Component::Inventory>(entity);
+        const auto& boots = inventory.boots();
+        if(boots) {
```
Comment (L3nn0x) [inline src/map/src/utils/calculation.cpp]: replace those if with the getter in utils/getters.h

File: unix_env_setup.sh [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 sudo apt-add-repository -y "ppa:ubuntu-toolchain-r/test"
 sudo apt-get update
-sudo apt-get install g++-8 unzip autoconf libtool libssl-dev python perl libncurses5-dev libreadline-dev
+sudo apt-get install g++-8 unzip autoconf libtool libssl-dev python libncurses5-dev libreadline-dev
```
Comment (L3nn0x) [inline unix_env_setup.sh:5]: g++-9 instead of g++-8 no?

Comment (RavenX8) [reply unix_env_setup.sh:5]: no, I maintain this to match what we are using on travis. Travis uses g++-8

File: src/map/src/utils/calculation.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    return attackSpeed;
+  }
+
+  int get_weight(EntitySystem& entitySystem, RoseCommon::Entity entity) {
+    return 0;
+  }
+
+  int get_maxhp(EntitySystem& entitySystem, RoseCommon::Entity entity) {
+    auto& life = entitySystem.get_component<Component::Life>(entity);
+    //TODO do calc here
```
Comment (L3nn0x) [inline src/map/src/utils/calculation.cpp:84]: which calc? It's life.mapHp

File: src/map/src/utils/calculation.cpp [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    return 0;
+  }
+
+  int get_maxmp(EntitySystem& entitySystem, RoseCommon::Entity entity) {
+    auto& magic = entitySystem.get_component<Component::Magic>(entity);
+    //TODO do calc here
```
Comment (L3nn0x) [inline src/map/src/utils/calculation.cpp:90]: which calc? It's magic.maxMp

File: src/map/src/utils/calculation.cpp (hunk 4 above)
Comment (RavenX8) [reply src/map/src/utils/calculation.cpp:84]: this is for calculating the max hp based on all stats and buffs. This isn't used yet

File: src/map/src/utils/calculation.cpp (hunk 5 above)
Comment (RavenX8) [reply src/map/src/utils/calculation.cpp:90]: this is for calculating the max mp based on all stats and buffs. This isn't used yet.

File: src/map/src/combat/combat.cpp [hunk 6]
```diff
@@ -215,56 +222,135 @@ void Combat::update(EntitySystem& entitySystem, Entity entity, uint32_t dt) {
...
+          auto exp_required = get_exp_to_level(attackerLevel.level+1);
+          while (attackerLevel.xp >= exp_required) {
+            int32_t new_exp = attackerLevel.xp - exp_required;
+            attackerLevel.xp = new_exp;
+            ++attackerLevel.level;
+            exp_required = get_exp_to_level(attackerLevel.level+1);
+
+            //TODO:: adjust our hp and mp based on our level here
+            //TODO:: give the player attr and skill points
+            auto p = SrvLevelup::create(attack_log.attacker_, attackerLevel.level, attackerLevel.xp, 0, 0);
```
Comment (L3nn0x) [inline src/map/src/combat/combat.cpp:297]: maybe have a dedicated function to take care of the whole level up thingy?

File: src/map/src/combat/combat.cpp [hunk 7]
```diff
@@ -215,56 +222,135 @@ void Combat::update(EntitySystem& entitySystem, Entity entity, uint32_t dt) {
...
+  // Check to see if we have a target component
   if(entitySystem.has_component<Component::Target>(entity) == true)
   {
     auto& target = entitySystem.get_component<Component::Target>(entity);
     //const auto& targetBasicInfo = entitySystem.get_component<Component::BasicInfo>(target.target);
-    const auto& targetLife = entitySystem.get_component<Component::Life>(target.target);
-    
+
     if(((entitySystem.has_component<Component::Npc>(target.target) == false && entitySystem.has_component<Component::Mob>(target.target) == true) ||
       (false))) // TODO:: Check if this map has PVP turned on and the target player isn't on my team
```
Comment (L3nn0x) [inline src/map/src/combat/combat.cpp:335]: maybe we want to have a function to check if the target is valid within a map

File: src/map/src/combat/combat.cpp [hunk 8]
```diff
@@ -418,6 +504,25 @@ void Combat::revive(EntitySystem& entitySystem, Entity entity, const RoseCommon:
   entitySystem.teleport_entity(entity, std::get<1>(dest), std::get<2>(dest), std::get<0>(dest));
 }
 
+int32_t Combat::get_exp_to_level(int level) {
```
Comment (L3nn0x) [inline src/map/src/combat/combat.cpp]: prob should be a int64_t here no?

**Review by L3nn0x — COMMENTED** [review]
looks ok

File: src/map/src/combat/combat.cpp (hunk 8 above)
Comment (RavenX8) [reply src/map/src/combat/combat.cpp]: the level component is int32_t, only made that match that.

Comment (RavenX8) [reply src/map/src/combat/combat.cpp]: Anyway I have changed it in the latest

File: src/map/src/combat/combat.cpp (hunk 6 above)
Comment (RavenX8) [reply src/map/src/combat/combat.cpp:297]: Will do this in next combat pass

File: src/map/src/combat/combat.cpp (hunk 7 above)
Comment (RavenX8) [reply src/map/src/combat/combat.cpp:335]: Will do this in next combat pass

**Review by L3nn0x — APPROVED** [review]

File: src/map/src/combat/combat.cpp (hunk 8 above)
Comment (L3nn0x) [reply src/map/src/combat/combat.cpp]: I had put in32_t here because the packets expect that 🤷‍♂

**Pull request merged**